controller/src/main/java/com/flipkart/varadhi/controller/impl/opexecutors/StartOpExecutor.java

```diff
@@ -0,0 +1,81 @@
+package com.flipkart.varadhi.controller.impl.opexecutors;
+
+import com.flipkart.varadhi.controller.AssignmentManager;
+import com.flipkart.varadhi.controller.OperationMgr;
+import com.flipkart.varadhi.core.cluster.ConsumerApi;
+import com.flipkart.varadhi.core.cluster.ConsumerClientFactory;
+import com.flipkart.varadhi.entities.SubscriptionShards;
+import com.flipkart.varadhi.entities.SubscriptionUnitShard;
+import com.flipkart.varadhi.entities.VaradhiSubscription;
+import com.flipkart.varadhi.entities.cluster.Assignment;
+import com.flipkart.varadhi.entities.cluster.OrderedOperation;
+import com.flipkart.varadhi.entities.cluster.ShardOperation;
+import com.flipkart.varadhi.entities.cluster.SubscriptionOperation;
+import com.flipkart.varadhi.spi.db.MetaStore;
+import lombok.extern.slf4j.Slf4j;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+
+@Slf4j
+public class StartOpExecutor extends SubscriptionStartShardExecutor {
+    public StartOpExecutor(
+            VaradhiSubscription subscription, ConsumerClientFactory clientFactory, OperationMgr operationMgr,
+            AssignmentManager assignmentManager, MetaStore metaStore
+    ) {
+        super(subscription, clientFactory, operationMgr, assignmentManager, metaStore);
+    }
+
+    @Override
+    public CompletableFuture<Void> execute(OrderedOperation operation) {
+        return startShards((SubscriptionOperation) operation, subscription);
+    }
+
+    private CompletableFuture<Void> startShards(SubscriptionOperation subOp, VaradhiSubscription subscription) {
+        SubscriptionShards shards = subscription.getShards();
+        return getOrCreateShardAssignment(subscription).thenCompose(assignments -> {
+            List<CompletableFuture<Boolean>> shardFutures = scheduleStartOnShards(subscription, subOp, assignments);
+            log.info("Executed Start on {} shards for SubOp({}).", shards.getShardCount(), subOp.getData());
+            return CompletableFuture.allOf(shardFutures.toArray(new CompletableFuture[0])).thenApply(ignore -> {
+                if (allShardsSkipped(shardFutures)) {
+                    log.info("Start {} completed without any shard operations being scheduled.", subOp.getData());
+                    completeSubOperation(subOp);
+                }
+                return null;
+            });
+        });
+    }
+
+    private CompletableFuture<List<Assignment>> getOrCreateShardAssignment(VaradhiSubscription subscription) {
+        List<SubscriptionUnitShard> unAssigned = getSubscriptionShards(subscription.getShards());
+        return assignmentManager.assignShards(unAssigned, subscription, List.of());
```

Comment:
Need to think about assignmentManager's interaction with operation's processing. I am assuming that all operation's processing can happen concurrently, so how is assignment manager handling this concurrent "assignment" requests.

Reply:
Operation for a subscriptions are not concurrent but across subscriptions they are. This is taken care in operation mgr itself.
Assignment mgr deals with shard (re|un)assignment flow. It is a single threaded executor which ensures no assignments are executed in parallel.

Reply:
Sure will go through that.